### PR TITLE
Fix something in ReplicatedMergeTree

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

 - Fixed some trash around REPLACE/MOVE PARTITION (fix wrong block numbers; create empty covering part on initiator too to avoid unexpected parts on restart). Follow-up to https://github.com/ClickHouse/ClickHouse/pull/56282
 - More correct (but less optimal) handling of unexpected Outdated parts, follow-up to https://github.com/ClickHouse/ClickHouse/pull/56693 
 - Fixed "The specified key doesn't exist" for empty parts. Was broken in https://github.com/ClickHouse/ClickHouse/pull/56282. Caught by `00993_system_parts_race_condition_drop_zookeeper` ([example](https://s3.amazonaws.com/clickhouse-test-reports/56552/ff11d69be4ffcba3970b0c3711aceb7e40baba09/stateless_tests__release__s3_storage__[2_2].html))

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit ef3e558a65715357cad047b8ae35eaf18b2013c5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57129/ef3e558a65715357cad047b8ae35eaf18b2013c5/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6962213873">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57129/ef3e558a65715357cad047b8ae35eaf18b2013c5/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by tavplubix** [comment]
Integration tests (asan) [1/4] - https://github.com/ClickHouse/ClickHouse/issues/56627
Integration tests (asan) [4/4]  - https://github.com/ClickHouse/ClickHouse/issues/55168
Stateless tests (debug) [2/5] - https://github.com/ClickHouse/ClickHouse/issues/49699

File: src/Storages/MergeTree/IMergeTreeDataPart.h [hunk 1]
```diff
@@ -246,6 +246,8 @@ class IMergeTreeDataPart : public std::enable_shared_from_this<IMergeTreeDataPar
         REMOVE_BLOBS,
         /// is set when Clickhouse is sure that the blobs belong to other replica and current replica has not locked them on s3 yet
         PRESERVE_BLOBS,
+        /// remove blobs even if the part is not temporary
+        REMOVE_BLOBS_OF_NOT_TEMPORARY,
```
Comment (CheSema) [inline src/Storages/MergeTree/IMergeTreeDataPart.h:250]: What do you think if we rename all enum cases?

REMOVE_BLOBS -> REMOVE_BLOBS_WHEN_TEMPORARY_STATE
PRESERVE_BLOBS -> PRESERVE_BLOBS_WHEN_TEMPORARY_STATE
REMOVE_BLOBS_OF_NOT_TEMPORARY -> REMOVE_BLOBS_WHEN_ANY_STATE

But I do not insist at all. May be it is ok but in separate PR, because that is the fix.

Comment (tavplubix) [reply src/Storages/MergeTree/IMergeTreeDataPart.h:250]: I can do it in a separate PR, but I don't really care. The enum name (`BlobsRemovalPolicyForTemporaryParts`) implies that it's for temporary parts. Maybe I should not have added the new value to this enum (because the new value is used for non-temporary parts), but I didn't want to introduce another special bool flag in `IMergeTreeDataPart`

**Review by CheSema — APPROVED** [review]
